packages/playwright-core/src/server/input.ts

```diff
@@ -44,40 +50,66 @@ export class Keyboard {
   private _pressedKeys = new Set<string>();
   private _raw: RawKeyboard;
   private _page: Page;
+  private _layoutName: string;
+  private _keyboardLayout: KeyboardLayoutClosure;
+  private _deadKeyMappings?: Map<string, string>;
 
   constructor(raw: RawKeyboard, page: Page) {
     this._raw = raw;
     this._page = page;
+    const layoutName = page._browserContext._options.keyboardLayout;
+    this._keyboardLayout = getByKeyboardLayoutName(layoutName);
+    this._layoutName = layoutName ?? 'us';
+  }
+
+  changeLayout(layoutName: string) {
+    this._keyboardLayout = getByKeyboardLayoutName(layoutName);
+    this._layoutName = layoutName;
+    this._deadKeyMappings = undefined;
   }
 
   async down(key: string) {
     const description = this._keyDescriptionForString(key);
+    if (description.key !== 'Shift') this._deadKeyMappings = undefined;
     const autoRepeat = this._pressedKeys.has(description.code);
     this._pressedKeys.add(description.code);
     if (kModifiers.includes(description.key as types.KeyboardModifier))
       this._pressedModifiers.add(description.key as types.KeyboardModifier);
-    const text = description.text;
-    await this._raw.keydown(this._pressedModifiers, description.code, description.keyCode, description.keyCodeWithoutLocation, description.key, description.location, autoRepeat, text);
+    await this._raw.keydown(this._pressedModifiers, description.code, description.keyCode, description.keyCodeWithoutLocation, description.key, description.location, autoRepeat, description.text);
   }
 
   private _keyDescriptionForString(keyString: string): KeyDescription {
-    let description = usKeyboardLayout.get(keyString);
+    let description = this._keyboardLayout.get(keyString);
     assert(description, `Unknown key: "${keyString}"`);
+    assert(!Array.isArray(description), `Accented key "${keyString}" cannot be typed with a single keypress on "${normalizeLayoutName(this._layoutName)}" keyboard layout.\nUse type() method that will generate multiple key presses.`);
+
     const shift = this._pressedModifiers.has('Shift');
     description = shift && description.shifted ? description.shifted : description;
 
     // if any modifiers besides shift are pressed, no text should be sent
     if (this._pressedModifiers.size > 1 || (!this._pressedModifiers.has('Shift') && this._pressedModifiers.size === 1))
       return { ...description, text: '' };
-    return description;
+
+    if (!this._deadKeyMappings || description.key === 'Shift') return description;
```

Comment:
Why special-casing `Shift` here? I guess this should be the same list as in `down()` method?

Reply:
That's correct.